massa-pos-exports/src/cycle_info.rs

```diff
@@ -0,0 +1,321 @@
+    /// Check if the production stats are above the required percentage
+    pub fn is_satisfying(&self, max_miss_ratio: &Ratio<u64>) -> bool {
+        let opportunities_count = self.block_success_count + self.block_failure_count;
+        if opportunities_count == 0 {
+            return true;
+        }
+        &Ratio::new(self.block_failure_count, opportunities_count) <= max_miss_ratio
+    }
+
+    /// Increment a production stat structure with another
+    pub fn extend(&mut self, stats: &ProductionStats) {
+        self.block_success_count = self
+            .block_success_count
+            .saturating_add(stats.block_success_count);
+        self.block_failure_count = self
+            .block_failure_count
+            .saturating_add(stats.block_failure_count);
+    }
+}
+
+/// Serializer for `ProductionStats`
+pub struct ProductionStatsSerializer {
+    u64_ser: U64VarIntSerializer,
+}
+
+impl Default for ProductionStatsSerializer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ProductionStatsSerializer {
+    /// Creates a new `ProductionStats` serializer
+    pub fn new() -> Self {
+        Self {
+            u64_ser: U64VarIntSerializer::new(),
+        }
+    }
+}
+
+impl Serializer<PreHashMap<Address, ProductionStats>> for ProductionStatsSerializer {
+    fn serialize(
+        &self,
+        value: &PreHashMap<Address, ProductionStats>,
+        buffer: &mut Vec<u8>,
+    ) -> Result<(), SerializeError> {
+        self.u64_ser.serialize(&(value.len() as u64), buffer)?;
+        for (
+            addr,
+            ProductionStats {
+                block_success_count,
+                block_failure_count,
+            },
+        ) in value.iter()
+        {
+            buffer.extend(addr.to_bytes());
+            self.u64_ser.serialize(block_success_count, buffer)?;
+            self.u64_ser.serialize(block_failure_count, buffer)?;
+        }
+        Ok(())
+    }
+}
+
+/// Deserializer for `ProductionStats`
+pub struct ProductionStatsDeserializer {
+    address_deserializer: AddressDeserializer,
+    u64_deserializer: U64VarIntDeserializer,
+}
+
+impl Default for ProductionStatsDeserializer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ProductionStatsDeserializer {
+    /// Creates a new `ProductionStats` deserializer
+    pub fn new() -> ProductionStatsDeserializer {
+        ProductionStatsDeserializer {
+            address_deserializer: AddressDeserializer::new(),
+            u64_deserializer: U64VarIntDeserializer::new(Included(u64::MIN), Included(u64::MAX)),
+        }
+    }
+}
+
+impl Deserializer<PreHashMap<Address, ProductionStats>> for ProductionStatsDeserializer {
+    fn deserialize<'a, E: ParseError<&'a [u8]> + ContextError<&'a [u8]>>(
+        &self,
+        buffer: &'a [u8],
+    ) -> IResult<&'a [u8], PreHashMap<Address, ProductionStats>, E> {
+        context(
+            "Failed ProductionStats deserialization",
+            length_count(
```

Comment:
A maximum value value should be set for this deserializer to avoid allocating big vector